src/firefox.py

```diff
@@ -0,0 +1,135 @@
+VERSION_REGEX = r"\d+(\.\d+)*"
+
+class UnsupportedReleasePageError(Exception):
+    "Raised when a firefox release page is not supported"
+    pass
+
+def format_date(unformatted_date: str) -> str:
+    """ Format date from July 11, 2002 to 2002-07-11 """
+    date = re.sub(r'(\d)(st|nd|rd|th)', r'\1', unformatted_date)
+    formats = ["%b %d, %Y", "%B %d, %Y"]
+    for f in formats:
+        try:
+            return datetime.strptime(date, f).strftime("%Y-%m-%d")
+        except ValueError:
+            pass
+    return ""
+
+def get_version_and_date_gt_28(soup: BeautifulSoup) -> Tuple[str, str]:
+    """ Version matching for firefox versions >= 28.0 """
+
+    # get version
+    version = soup.find("div", class_="c-release-version").get_text()
+
+    # get date
+    unformatted_date = soup.find("p", class_="c-release-date").get_text()
+    date = format_date(unformatted_date)
+
+    return (version, date)
+
+def get_version_and_date_gt_10(soup: BeautifulSoup) -> Tuple[str, str]:
+    """ Version matching for firefox versions >= 10.0 """
+    release_info = soup.find("h2").find("small").text
+
+    # get version
+    version_match = re.search(VERSION_REGEX, soup.select('div#nav-access a')[0].get("href"))
+    if version_match is None:
+        raise UnsupportedReleasePageError("Unable to find version")
+    version = version_match.group()
+
+    # get date
+    unformatted_date_match = re.search(DATE_REGEX, release_info)
+    if unformatted_date_match is None:
+        raise UnsupportedReleasePageError("Unable to find date")
+    unformatted_date = unformatted_date_match.group()
+    date = format_date(unformatted_date)
+
+    return (version, date)
+
+def get_version_and_date_gt_3(soup: BeautifulSoup) -> Tuple[str, str]:
+    """ Version matching for firefox versions >= 3.0 """
+    release_info = soup.select('div#main-feature p em')[0].get_text()
+
+    # get version
+    version_match = re.search(VERSION_REGEX, release_info)
+    if version_match is None:
+        raise UnsupportedReleasePageError("Unable to find version")
+    version = version_match.group()
+
+    # get date
+    unformatted_date_match = re.search(DATE_REGEX, release_info)
+    if unformatted_date_match is None:
+        raise UnsupportedReleasePageError("Unable to find date")
+    unformatted_date = unformatted_date_match.group()
+    date = format_date(unformatted_date)
+
+    return (version, date)
+
+def get_version_and_date(soup: BeautifulSoup, release_page: str) -> Tuple[str, str]:
+    """ Get version and date from the given release page """
+    functions = [get_version_and_date_gt_28, get_version_and_date_gt_10, get_version_and_date_gt_3]
+
+    # Note: firefox release pages for versions <3.0 don't include release dates so we
+    # can't match these versions for now.
+    # example: https://www.mozilla.org/en-US/firefox/2.0/releasenotes/
+    for function in functions:
```

Comment:
If I understood correctly, here we know the version. It would be best to call the right function directly to avoid unnecessarily requests and reduce the execution time.

Reply:
Upon further investigation, it seems the pages come in 3 variants and _mostly_ follow version numbers, but not exactly. So I think we will need to keep this try/except checking. Added more comments in the code about this.

Reply:
> So I think we will need to keep this try/except checking.

Ok, let's keep it like that.



> I added some concurrency with python's `concurrent.futures` that should speed things up.

Indeed, that's way more faster. Hope there will be no issue due to the number of requests (I already got a timeout).

Reply:
There's no rate limiting on the www.mozilla.org page from what i've seen, but I have noticed timeouts on occasion (both when making concurrent requests and when not)

Im not sure where/how these python release jobs are run. Do we retry failed runs? If so, shouldn't be an issue. But I could also do timeout/exponentional backoff/retry handling here if we need.